### PR TITLE
Aggregate DVs conditions on VMI (and so VM)

### PR DESCRIPTION
**Pull request by tiraboschi**
**What this PR does / why we need it**:

Today we have really detailed conditions
on DataVolumes but they are not directly reported on
VMs and VMIs.
Conditions provide a standard mechanism for higher-level
status reporting from a controller:
when a VM/VMI is not in the expect state the user is
supposed to check its conditions to understand what's wrong.
Let's aggregate DataVolumes conditions in a top level one
("DataVolumesReady") to simple expose to the user
an aggregated top-level condition on VM/VMI summarizing
more detailed conditions from datavolumes.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes None

**Special notes for your reviewer**:
None

**Checklist**

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [X] PR: The PR description is expressive enough and will help future contributors
- [X] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [X] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [X] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

**Release note**:

```release-note
Aggregate DVs conditions on VMI (and so VM)
```

### CODE REVIEW

File: pkg/virt-controller/watch/vmi.go [hunk 1]
```diff
@@ -2399,3 +2401,97 @@ func (c *VMIController) syncMemoryHotplug(vmi *virtv1.VirtualMachineInstance) {
...
+
+	dvsReadyCondition := virtv1.VirtualMachineInstanceCondition{
+		Type:   virtv1.VirtualMachineInstanceDataVolumesReady,
+		Status: k8sv1.ConditionUnknown,
+	}
+
+	if allDVConditions[cdiv1.DataVolumeBound].Status == k8sv1.ConditionFalse {
+		dvsReadyCondition.Status = k8sv1.ConditionFalse
+		dvsReadyCondition.Reason = virtv1.VirtualMachineInstanceReasonNotAllDVsBound
+		dvsReadyCondition.Message = "Not all the VMI' DVs are bound: " + strings.Join(allDVConditions[cdiv1.DataVolumeBound].Message, ",")
```
Comment (0xFelix) [inline pkg/virt-controller/watch/vmi.go:2455]:
```suggestion
		dvsReadyCondition.Message = "Not all of the VMI's DVs are bound: " + strings.Join(allDVConditions[cdiv1.DataVolumeBound].Message, ",")
```

File: pkg/virt-controller/watch/vmi.go [hunk 2]
```diff
@@ -2399,3 +2401,97 @@ func (c *VMIController) syncMemoryHotplug(vmi *virtv1.VirtualMachineInstance) {
...
+	} else if allDVConditions[cdiv1.DataVolumeRunning].Status == k8sv1.ConditionTrue {
+		dvsReadyCondition.Status = k8sv1.ConditionFalse
+		dvsReadyCondition.Reason = virtv1.VirtualMachineInstanceReasonDVRunning
+		dvsReadyCondition.Message = "At least one of VMI' DVs is still running (import/upload/clone): " + strings.Join(allDVConditions[cdiv1.DataVolumeRunning].Message, ",")
```
Comment (0xFelix) [inline pkg/virt-controller/watch/vmi.go:2459]:
```suggestion
		dvsReadyCondition.Message = "At least one of the VMI's DVs is still running (import/upload/clone): " + strings.Join(allDVConditions[cdiv1.DataVolumeRunning].Message, ",")
```

File: pkg/virt-controller/watch/vmi.go [hunk 3]
```diff
@@ -2399,3 +2401,97 @@ func (c *VMIController) syncMemoryHotplug(vmi *virtv1.VirtualMachineInstance) {
...
+	} else if allDVConditions[cdiv1.DataVolumeReady].Status == k8sv1.ConditionTrue {
+		dvsReadyCondition.Status = k8sv1.ConditionTrue
+		dvsReadyCondition.Reason = virtv1.VirtualMachineInstanceReasonAllDVsReady
+		dvsReadyCondition.Message = "All the VMI' DVs are bound and not running"
```
Comment (0xFelix) [inline pkg/virt-controller/watch/vmi.go:2463]:
```suggestion
		dvsReadyCondition.Message = "All of the VMI's DVs are bound and not running"
```

File: pkg/virt-controller/watch/vmi_test.go [hunk 4]
```diff
@@ -212,12 +212,30 @@ var _ = Describe("VirtualMachineInstance watcher", func() {
...
 		vmiInterface.EXPECT().Update(context.Background(), gomock.Any()).Do(func(ctx context.Context, arg interface{}) {
 			Expect(arg.(*virtv1.VirtualMachineInstance).Status.Phase).To(Equal(virtv1.Failed))
-			Expect(arg.(*virtv1.VirtualMachineInstance).Status.Conditions).To(ConsistOf(MatchFields(IgnoreExtras,
+			Expect(arg.(*virtv1.VirtualMachineInstance).Status.Conditions).To(ContainElement(MatchFields(IgnoreExtras,
 				Fields{"Type": Equal(virtv1.VirtualMachineInstanceReady)})))
 			Expect(arg.(*virtv1.VirtualMachineInstance).Status.PhaseTransitionTimestamps).ToNot(BeEmpty())
 			Expect(arg.(*virtv1.VirtualMachineInstance).Status.PhaseTransitionTimestamps).ToNot(HaveLen(len(vmi.Status.PhaseTransitionTimestamps)))
 		}).Return(vmi, nil)
 	}
+	shouldExpectVirtualMachineDataVolumesReadyUnknown := func(vmi *virtv1.VirtualMachineInstance) {
```
Comment (0xFelix) [inline pkg/virt-controller/watch/vmi_test.go:221]: nit: Possibly dedup these funcs?

File: staging/src/kubevirt.io/api/core/v1/types.go [hunk 5]
```diff
@@ -523,6 +523,16 @@ const (
...
+
+	// Summarizes that all the DataVolumes attached to the VMI are Ready or not (at least one stil not Bound or still Running=import/upload/clone)
+	VirtualMachineInstanceDataVolumesReady = "DataVolumesReady"
+
+	// Reason means that not all the VMI' DV are already bound
+	VirtualMachineInstanceReasonNotAllDVsBound = "NotAllDVsBound"
+	// Reason means that at least one of the VMI' DV is still Running (import/upload/clone)
+	VirtualMachineInstanceReasonDVRunning = "DVRunning"
+	// Reason means that all the VMI' DVs are ready
+	VirtualMachineInstanceReasonAllDVsReady = "AllDVsReady"
```
Comment (0xFelix) [inline staging/src/kubevirt.io/api/core/v1/types.go:535]: Same as above for comments

File: tests/storage/datavolume.go [hunk 6]
```diff
@@ -381,6 +382,92 @@ var _ = SIGDescribe("DataVolume Integration", func() {
 					return vm.Status.PrintableStatus
 				}, 180*time.Second, 2*time.Second).Should(Equal(v1.VirtualMachineStatusStopped))
 			})
+
+			It("should accurately aggregate DataVolume conditions from many DVs", func() {
+				sc, err := libstorage.GetSnapshotStorageClass(virtClient)
+				if err != nil {
+					Skip("no snapshot storage class configured")
```
Comment (0xFelix) [inline tests/storage/datavolume.go:389]: Can we do it without `Skip`?

File: tests/storage/datavolume.go [hunk 7]
```diff
@@ -381,6 +382,92 @@ var _ = SIGDescribe("DataVolume Integration", func() {
...
+				By("requiring an additional DataVolume")
+				diskName := "disk1"
+				vmi = libstorage.AddDataVolumeDisk(vmi, diskName, dataVolume2.Name)
+
+				vmSpec := tests.NewRandomVirtualMachine(vmi, true)
+
+				vm, err := virtClient.VirtualMachine(vmSpec.Namespace).Create(context.Background(), vmSpec)
+				Expect(err).ToNot(HaveOccurred())
+
+				Eventually(func() v1.VirtualMachinePrintableStatus {
```
Comment (0xFelix) [inline tests/storage/datavolume.go:412]: There is a matcher for the printable status in `vm_test.go`. Maybe you could use `ThisVM` and the `printableStatus` matcher here too?

File: tests/storage/datavolume.go [hunk 8]
```diff
@@ -381,6 +382,92 @@ var _ = SIGDescribe("DataVolume Integration", func() {
...
+					return vm.Status.PrintableStatus
+				}, 180*time.Second, 2*time.Second).Should(Equal(v1.VirtualMachineStatusPvcNotFound))
+
+				By("creating the first DataVolume")
+				dataVolume1, err = virtClient.CdiClient().CdiV1beta1().DataVolumes(vmSpec.Namespace).Create(context.Background(), dataVolume1, metav1.CreateOptions{})
+				Expect(err).ToNot(HaveOccurred())
+				defer libstorage.DeleteDataVolume(&dataVolume1)
+
+				By("ensuring that VMI and VM are reporting VirtualMachineInstanceDataVolumesReady=False")
+				Eventually(func() []v1.VirtualMachineInstanceCondition {
```
Comment (0xFelix) [inline tests/storage/datavolume.go:424]: Can you dedup the `Eventually` blocks?

**Review by 0xFelix — COMMENTED** [review]
Added some comments, mostly nits.

File: tests/storage/datavolume.go (hunk 7 above)
Comment (tiraboschi) [reply tests/storage/datavolume.go]: done

File: tests/storage/datavolume.go (hunk 8 above)
Comment (tiraboschi) [reply tests/storage/datavolume.go]: done

File: pkg/virt-controller/watch/vmi.go [hunk 9]
```diff
@@ -2399,3 +2401,97 @@ func (c *VMIController) syncMemoryHotplug(vmi *virtv1.VirtualMachineInstance) {
...
+		dvsReadyCondition.Message = "All of the VMI's DVs are bound and not running"
+	}
+
+	vmiConditions := controller.NewVirtualMachineInstanceConditionManager()
+	vmiConditions.UpdateCondition(vmiCopy, &dvsReadyCondition)
+
+}
+
+func aggregateCondsFalseWins(condition cdiv1.DataVolumeCondition, allDVConditions map[cdiv1.DataVolumeConditionType]aggregateDataVolumeCondition) {
+	if condition.Status == k8sv1.ConditionFalse || condition.Status == k8sv1.ConditionUnknown {
```
Comment (0xFelix) [inline pkg/virt-controller/watch/vmi.go:2472]: Won't you lose the `condition.Reason` / `condition.Message` if this equates to `false`?

Wouldn't something like this be cleaner?

```
	status := allDVConditions[condition.Type].Status
	if (condition.Status == k8sv1.ConditionFalse || condition.Status == k8sv1.ConditionUnknown) &&
		allDVConditions[condition.Type].Status == k8sv1.ConditionTrue {
		status = condition.Status
	}
	allDVConditions[condition.Type] = aggregateDataVolumeCondition{
		Status:  status,
		Reason:  append(allDVConditions[condition.Type].Reason, condition.Reason),
		Message: append(allDVConditions[condition.Type].Reason, condition.Message),
	}
```

File: pkg/virt-controller/watch/vmi.go [hunk 10]
```diff
@@ -2399,3 +2401,97 @@ func (c *VMIController) syncMemoryHotplug(vmi *virtv1.VirtualMachineInstance) {
...
+		var status k8sv1.ConditionStatus
+		if allDVConditions[condition.Type].Status == k8sv1.ConditionTrue {
+			status = condition.Status
+		}
+		allDVConditions[condition.Type] = aggregateDataVolumeCondition{
+			Status:  status,
```
Comment (0xFelix) [inline pkg/virt-controller/watch/vmi.go:2478]: `status` is uninitialized when `allDVConditions[condition.Type].Status != k8sv1.ConditionTrue`?

File: pkg/virt-controller/watch/vmi.go [hunk 11]
```diff
@@ -2399,3 +2401,97 @@ func (c *VMIController) syncMemoryHotplug(vmi *virtv1.VirtualMachineInstance) {
...
+			Reason:  append(allDVConditions[condition.Type].Reason, condition.Reason),
+			Message: append(allDVConditions[condition.Type].Reason, condition.Message),
+		}
+	}
+}
+
+func aggregateCondsTrueWins(condition cdiv1.DataVolumeCondition, allDVConditions map[cdiv1.DataVolumeConditionType]aggregateDataVolumeCondition) {
+	if condition.Status == k8sv1.ConditionTrue || condition.Status == k8sv1.ConditionUnknown {
```
Comment (0xFelix) [inline pkg/virt-controller/watch/vmi.go:2486]: Same as above.

File: pkg/virt-controller/watch/vmi.go [hunk 12]
```diff
@@ -2399,3 +2401,97 @@ func (c *VMIController) syncMemoryHotplug(vmi *virtv1.VirtualMachineInstance) {
...
+		var status k8sv1.ConditionStatus
+		if allDVConditions[condition.Type].Status == k8sv1.ConditionFalse {
+			status = condition.Status
```
Comment (0xFelix) [inline pkg/virt-controller/watch/vmi.go:2489]: Same as above.

File: pkg/virt-controller/watch/vmi.go (hunk 10 above)
Comment (tiraboschi) [reply pkg/virt-controller/watch/vmi.go:2490]: Right, this was somehow masked by the set of if-s starting on line 2452 but let's explicitly avoid it with a safe initialization.
Let's also enlarge the test matrix in the unit tests.

File: pkg/virt-controller/watch/vmi.go (hunk 11 above)
Comment (tiraboschi) [reply pkg/virt-controller/watch/vmi.go:2498]: same

File: pkg/virt-controller/watch/vmi.go (hunk 12 above)
Comment (tiraboschi) [reply pkg/virt-controller/watch/vmi.go]: same

File: pkg/virt-controller/watch/vmi.go [hunk 13]
```diff
@@ -2399,3 +2401,109 @@ func (c *VMIController) syncMemoryHotplug(vmi *virtv1.VirtualMachineInstance) {
...
+			Status: ConditionUninitialized,
+		},
+		cdiv1.DataVolumeRunning: {
+			Status: ConditionUninitialized,
+		},
+	}
+
+	for _, dv := range dvs {
+		for _, condition := range dv.Status.Conditions {
+			if allDVConditions[condition.Type].Status == ConditionUninitialized {
```
Comment (0xFelix) [inline pkg/virt-controller/watch/vmi.go:2431]: Can't you do something like this instead of introducing the `ConditionUninitialized` const?

```
if _, ok := allDVConditions[condition.Type]; !ok {
	allDVConditions[condition.Type] = aggregateDataVolumeCondition{
		Status:  condition.Status,
		Reason:  []string{condition.Reason},
		Message: []string{condition.Message},
	}
} else {
	if condition.Type == cdiv1.DataVolumeRunning {
		aggregateCondsTrueWins(condition, allDVConditions)
	} else {
		aggregateCondsFalseWins(condition, allDVConditions)
	}
}
```

Comment (0xFelix) [reply pkg/virt-controller/watch/vmi.go:2431]: This should work with your status checks below too, because the returned default value should still satisfy the checks (default `Status` is empty).

File: pkg/virt-controller/watch/vmi.go [hunk 14]
```diff
@@ -2399,3 +2401,109 @@ func (c *VMIController) syncMemoryHotplug(vmi *virtv1.VirtualMachineInstance) {
...
+			}
+		}
+	}
+
+	dvsReadyCondition := virtv1.VirtualMachineInstanceCondition{
+		Type:   virtv1.VirtualMachineInstanceDataVolumesReady,
+		Status: k8sv1.ConditionUnknown,
+	}
+
+	if allDVConditions[cdiv1.DataVolumeBound].Status == k8sv1.ConditionFalse {
```
Comment (0xFelix) [inline pkg/virt-controller/watch/vmi.go:2452]: nit: Maybe make it a switch?

File: pkg/virt-controller/watch/vmi.go (hunk 9 above)
Comment (0xFelix) [reply pkg/virt-controller/watch/vmi.go:2484]: ACK

File: pkg/virt-controller/watch/vmi.go (hunk 13 above)
Comment (tiraboschi) [reply pkg/virt-controller/watch/vmi.go]: Right, I can simply consider an empty status as uninitialized.
Done

File: pkg/virt-controller/watch/vmi.go (hunk 14 above)
Comment (tiraboschi) [reply pkg/virt-controller/watch/vmi.go:2445]: Unfortunately I'm checking different values on the various branches.

File: pkg/virt-controller/watch/vmi.go [hunk 15]
```diff
@@ -2399,3 +2401,98 @@ func (c *VMIController) syncMemoryHotplug(vmi *virtv1.VirtualMachineInstance) {
...
+		dvsReadyCondition.Message = "Not all of the VMI's DVs are bound and not running (unknown)"
+	} else if allDVConditions[cdiv1.DataVolumeReady].Status == k8sv1.ConditionTrue {
+		dvsReadyCondition.Status = k8sv1.ConditionTrue
+		dvsReadyCondition.Reason = virtv1.VirtualMachineInstanceReasonAllDVsReady
+		dvsReadyCondition.Message = "All of the VMI's DVs are bound and not running"
+	}
+
+	vmiConditions := controller.NewVirtualMachineInstanceConditionManager()
+	vmiConditions.UpdateCondition(vmiCopy, &dvsReadyCondition)
+
```
Comment (0xFelix) [inline pkg/virt-controller/watch/vmi.go:2469]: nit: newline

**Review by 0xFelix — APPROVED** [review]
Thanks! Just one more nit, feel free to fix though.

/lgtm

File: pkg/virt-controller/watch/vmi.go (hunk 15 above)
Comment (tiraboschi) [reply pkg/virt-controller/watch/vmi.go]: done

**Review by 0xFelix — APPROVED** [review]
/lgtm

File: pkg/virt-controller/watch/vmi.go [hunk 16]
```diff
@@ -2399,3 +2401,97 @@ func (c *VMIController) syncMemoryHotplug(vmi *virtv1.VirtualMachineInstance) {
...
 	}
 }
+
+type aggregateDataVolumeCondition struct {
+	Status  k8sv1.ConditionStatus
+	Reason  []string
+	Message []string
+}
+
+func (c *VMIController) aggregateDataVolumesConditions(vmiCopy *virtv1.VirtualMachineInstance, dvs []*cdiv1.DataVolume) {
```
Comment (xpivarc) [inline pkg/virt-controller/watch/vmi.go:2411]: Do we need to recompute this once the VMI is running?

Comment (acardace) [inline pkg/virt-controller/watch/vmi.go:2411]:
```suggestion
func (c *VMIController) aggregateDataVolumesConditions(vmiCopy *virtv1.VirtualMachineInstance, dvs []*cdiv1.DataVolume) {
	dvsReadyCondition := virtv1.VirtualMachineInstanceCondition{
		Status: k8sv1.ConditionTrue,
		Type:   virtv1.VirtualMachineInstanceDataVolumesReady,
		Reason: virtv1.VirtualMachineInstanceReasonAllDVsReady,
		Message: "All of the VMI's DVs are bound and not running",
	}

	for _, dv := range dvs {
		if dv.Status.Phase != cdiv1.Succeeded { 
			dvsReadyCondition.Message = "Not all of the VMI's DVs are ready"
			dvsReadyCondition.Status = k8sv1.ConditionFalse
			break
		}
	}

	vmiConditions := controller.NewVirtualMachineInstanceConditionManager()
	vmiConditions.UpdateCondition(vmiCopy, &dvsReadyCondition)
}
```

**Review by acardace — CHANGES_REQUESTED** [review]
@tiraboschi I think the PR is really useful as it address the issue that the user has no indication when something is wrong with data volumes but I find the implementation to be a bit problematic on a few things.

First I think it relies too much on the internal working of CDI, I know KubeVirt is already coded in a way that it relies a lot on CDI but I find that this code makes too many assumptions about which conditions exist and in which order are populated. 

I think the code tries too hard to give too many details that are not so much of interest to a user, specifically, I don't think it's worth giving out so much info about why a data volume is not ready, what's interesting is if a VM is not able to start because a DV is not running so that a user is pointed to where the problem is. The problematic DVs exist in the same namespace so if needed someone can inspect those to get the right level of detail, we don't need KubeVirt to report so much about an object managed by a different operator, also because inner working may change in CDI and this code would get broken without us noticing.

The last point is a consequence of what I just wrote above, I think the code ended up being a bit too convoluted and hard to follow because of the previous point.

My suggestion would be to simplify all of this by just reporting whether the DVs are ready or not in the newly added condition, I suggested a snippet in a comment.

File: pkg/virt-controller/watch/vmi.go (hunk 16 above)
Comment (tiraboschi) [reply pkg/virt-controller/watch/vmi.go:2411]: > 		if dv.Status.Phase != cdiv1.Succeeded { 

According to [K8s API Conventions](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#typical-status-properties), we should explicitly monitor the individual conditions and do not implicitly infer from phases.

See:

> Some resources in the v1 API contain fields called phase, and associated message, reason, and other status fields. **The pattern of using phase is deprecated. Newer API types should use conditions instead.** Phase was essentially a state-machine enumeration field, that contradicted [system-design principles](https://git.k8s.io/design-proposals-archive/architecture/principles.md#control-logic) and hampered evolution, since [adding new enum values breaks backward compatibility](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md). Rather than encouraging clients to infer implicit properties from phases, we prefer to explicitly expose the individual conditions that clients need to monitor. Conditions also have the benefit that it is possible to create some conditions with uniform meaning across all resource types, while still exposing others that are unique to specific resource types. See [#7856](http://issues.k8s.io/7856) for more details and discussion.

Comment (acardace) [reply pkg/virt-controller/watch/vmi.go:2411]: I'm okay with monitoring the conditions only, but I think we should try to fail fast and early and try not to replace the info that CDI already exposes. This would simplify the code a lot.

Comment (tiraboschi) [reply pkg/virt-controller/watch/vmi.go:2411]: I pushed a simplified version (looking only at cdiv1.DataVolumeReady conditions) at https://github.com/kubevirt/kubevirt/pull/10905

**Comment by tiraboschi** [comment]
Closing for https://github.com/kubevirt/kubevirt/pull/10905 which implements a simplified version of it.
